{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import AGEpy as age\n",
    "import copy\n",
    "from gtfparse import read_gtf\n",
    "import numpy as np\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import data and format as a gtf file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/sahil/anaconda3/envs/gtf_formatting/lib/python3.9/site-packages/gtfparse/read_gtf.py:151: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version.\n",
      "\n",
      "\n",
      "  result = parse_gtf(\n",
      "/Users/sahil/anaconda3/envs/gtf_formatting/lib/python3.9/site-packages/gtfparse/read_gtf.py:151: FutureWarning: The warn_bad_lines argument has been deprecated and will be removed in a future version.\n",
      "\n",
      "\n",
      "  result = parse_gtf(\n",
      "INFO:root:Extracted GTF attributes: ['gene_id', 'gene_name', 'transcript_id', 'tss_id', 'p_id']\n",
      "/var/folders/8_/t_24z32d4_32jn404jwdwdfm0000gn/T/ipykernel_12721/4171810911.py:4: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  genes_df_frames[\"frame\"] = \".\"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>tss_id</th>\n",
       "      <th>p_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>11874</td>\n",
       "      <td>12227</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>NR_046018</td>\n",
       "      <td>TSS14844</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75545</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>CDS</td>\n",
       "      <td>216497510</td>\n",
       "      <td>216497694</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>USH2A</td>\n",
       "      <td>USH2A</td>\n",
       "      <td>NM_206933</td>\n",
       "      <td>TSS27031</td>\n",
       "      <td>P4249</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75548</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>CDS</td>\n",
       "      <td>216498647</td>\n",
       "      <td>216498941</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>USH2A</td>\n",
       "      <td>USH2A</td>\n",
       "      <td>NM_007123</td>\n",
       "      <td>TSS27031</td>\n",
       "      <td>P5706</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75549</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>CDS</td>\n",
       "      <td>216498647</td>\n",
       "      <td>216498941</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>USH2A</td>\n",
       "      <td>USH2A</td>\n",
       "      <td>NM_206933</td>\n",
       "      <td>TSS27031</td>\n",
       "      <td>P4249</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75552</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>CDS</td>\n",
       "      <td>216500933</td>\n",
       "      <td>216500996</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>USH2A</td>\n",
       "      <td>USH2A</td>\n",
       "      <td>NM_007123</td>\n",
       "      <td>TSS27031</td>\n",
       "      <td>P5706</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>865010</th>\n",
       "      <td>chrY</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>15417918</td>\n",
       "      <td>15417992</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>UTY</td>\n",
       "      <td>UTY</td>\n",
       "      <td>NR_047619</td>\n",
       "      <td>TSS25449</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>865009</th>\n",
       "      <td>chrY</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>15417918</td>\n",
       "      <td>15417992</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>UTY</td>\n",
       "      <td>UTY</td>\n",
       "      <td>NM_001258258</td>\n",
       "      <td>TSS25449</td>\n",
       "      <td>P8418</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>865008</th>\n",
       "      <td>chrY</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>15417918</td>\n",
       "      <td>15417992</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>UTY</td>\n",
       "      <td>UTY</td>\n",
       "      <td>NR_047643</td>\n",
       "      <td>TSS25449</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>865014</th>\n",
       "      <td>chrY</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>15417918</td>\n",
       "      <td>15417992</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>UTY</td>\n",
       "      <td>UTY</td>\n",
       "      <td>NM_001258256</td>\n",
       "      <td>TSS25449</td>\n",
       "      <td>P29049</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>869199</th>\n",
       "      <td>chrY</td>\n",
       "      <td>unknown</td>\n",
       "      <td>CDS</td>\n",
       "      <td>59342487</td>\n",
       "      <td>59343077</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>IL9R</td>\n",
       "      <td>IL9R</td>\n",
       "      <td>NM_002186</td>\n",
       "      <td>TSS13438</td>\n",
       "      <td>P19376</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>869204 rows Ã— 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       seqname   source feature      start        end score strand frame  \\\n",
       "0         chr1  unknown    exon      11874      12227     .      +     .   \n",
       "75545     chr1  unknown     CDS  216497510  216497694     .      -     0   \n",
       "75548     chr1  unknown     CDS  216498647  216498941     .      -     1   \n",
       "75549     chr1  unknown     CDS  216498647  216498941     .      -     1   \n",
       "75552     chr1  unknown     CDS  216500933  216500996     .      -     2   \n",
       "...        ...      ...     ...        ...        ...   ...    ...   ...   \n",
       "865010    chrY  unknown    exon   15417918   15417992     .      -     .   \n",
       "865009    chrY  unknown    exon   15417918   15417992     .      -     .   \n",
       "865008    chrY  unknown    exon   15417918   15417992     .      -     .   \n",
       "865014    chrY  unknown    exon   15417918   15417992     .      -     .   \n",
       "869199    chrY  unknown     CDS   59342487   59343077     .      +     0   \n",
       "\n",
       "        gene_id gene_name transcript_id    tss_id    p_id  \n",
       "0       DDX11L1   DDX11L1     NR_046018  TSS14844          \n",
       "75545     USH2A     USH2A     NM_206933  TSS27031   P4249  \n",
       "75548     USH2A     USH2A     NM_007123  TSS27031   P5706  \n",
       "75549     USH2A     USH2A     NM_206933  TSS27031   P4249  \n",
       "75552     USH2A     USH2A     NM_007123  TSS27031   P5706  \n",
       "...         ...       ...           ...       ...     ...  \n",
       "865010      UTY       UTY     NR_047619  TSS25449          \n",
       "865009      UTY       UTY  NM_001258258  TSS25449   P8418  \n",
       "865008      UTY       UTY     NR_047643  TSS25449          \n",
       "865014      UTY       UTY  NM_001258256  TSS25449  P29049  \n",
       "869199     IL9R      IL9R     NM_002186  TSS13438  P19376  \n",
       "\n",
       "[869204 rows x 13 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Read in gtf file and format necessary columns\n",
    "genes_df = read_gtf(\"genes.gtf\")\n",
    "genes_df_frames = genes_df[genes_df[\"feature\"] != \"CDS\"]\n",
    "genes_df_frames[\"frame\"] = \".\"\n",
    "genes_df_cds = genes_df[genes_df[\"feature\"] == \"CDS\"]\n",
    "genes_df = pd.concat([genes_df_frames, genes_df_cds])\n",
    "genes_df = genes_df.sort_values(by=\"seqname\")\n",
    "genes_df[\"score\"] = \".\"\n",
    "genes_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>tss_id</th>\n",
       "      <th>p_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>11874</td>\n",
       "      <td>12227</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>NR_046018</td>\n",
       "      <td>TSS14844</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78313</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>228337415</td>\n",
       "      <td>228337708</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>GJC2</td>\n",
       "      <td>GJC2</td>\n",
       "      <td>NM_020435</td>\n",
       "      <td>TSS4470</td>\n",
       "      <td>P27430</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78307</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>228336366</td>\n",
       "      <td>228336655</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>GUK1</td>\n",
       "      <td>GUK1</td>\n",
       "      <td>NM_001242839</td>\n",
       "      <td>TSS23291</td>\n",
       "      <td>P24635</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78314</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>228345441</td>\n",
       "      <td>228347527</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>GJC2</td>\n",
       "      <td>GJC2</td>\n",
       "      <td>NM_020435</td>\n",
       "      <td>TSS4470</td>\n",
       "      <td>P27430</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78306</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>228336366</td>\n",
       "      <td>228336655</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>GUK1</td>\n",
       "      <td>GUK1</td>\n",
       "      <td>NM_001242840</td>\n",
       "      <td>TSS15089</td>\n",
       "      <td>P4110</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>865011</th>\n",
       "      <td>chrY</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>15417918</td>\n",
       "      <td>15417992</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>UTY</td>\n",
       "      <td>UTY</td>\n",
       "      <td>NR_047622</td>\n",
       "      <td>TSS25449</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>865010</th>\n",
       "      <td>chrY</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>15417918</td>\n",
       "      <td>15417992</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>UTY</td>\n",
       "      <td>UTY</td>\n",
       "      <td>NR_047619</td>\n",
       "      <td>TSS25449</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>865009</th>\n",
       "      <td>chrY</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>15417918</td>\n",
       "      <td>15417992</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>UTY</td>\n",
       "      <td>UTY</td>\n",
       "      <td>NM_001258258</td>\n",
       "      <td>TSS25449</td>\n",
       "      <td>P8418</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>865008</th>\n",
       "      <td>chrY</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>15417918</td>\n",
       "      <td>15417992</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>UTY</td>\n",
       "      <td>UTY</td>\n",
       "      <td>NR_047643</td>\n",
       "      <td>TSS25449</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>865014</th>\n",
       "      <td>chrY</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>15417918</td>\n",
       "      <td>15417992</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>UTY</td>\n",
       "      <td>UTY</td>\n",
       "      <td>NM_001258256</td>\n",
       "      <td>TSS25449</td>\n",
       "      <td>P29049</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>430178 rows Ã— 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       seqname   source feature      start        end score strand frame  \\\n",
       "0         chr1  unknown    exon      11874      12227     .      +     .   \n",
       "78313     chr1  unknown    exon  228337415  228337708     .      +     .   \n",
       "78307     chr1  unknown    exon  228336366  228336655     .      +     .   \n",
       "78314     chr1  unknown    exon  228345441  228347527     .      +     .   \n",
       "78306     chr1  unknown    exon  228336366  228336655     .      +     .   \n",
       "...        ...      ...     ...        ...        ...   ...    ...   ...   \n",
       "865011    chrY  unknown    exon   15417918   15417992     .      -     .   \n",
       "865010    chrY  unknown    exon   15417918   15417992     .      -     .   \n",
       "865009    chrY  unknown    exon   15417918   15417992     .      -     .   \n",
       "865008    chrY  unknown    exon   15417918   15417992     .      -     .   \n",
       "865014    chrY  unknown    exon   15417918   15417992     .      -     .   \n",
       "\n",
       "        gene_id gene_name transcript_id    tss_id    p_id  \n",
       "0       DDX11L1   DDX11L1     NR_046018  TSS14844          \n",
       "78313      GJC2      GJC2     NM_020435   TSS4470  P27430  \n",
       "78307      GUK1      GUK1  NM_001242839  TSS23291  P24635  \n",
       "78314      GJC2      GJC2     NM_020435   TSS4470  P27430  \n",
       "78306      GUK1      GUK1  NM_001242840  TSS15089   P4110  \n",
       "...         ...       ...           ...       ...     ...  \n",
       "865011      UTY       UTY     NR_047622  TSS25449          \n",
       "865010      UTY       UTY     NR_047619  TSS25449          \n",
       "865009      UTY       UTY  NM_001258258  TSS25449   P8418  \n",
       "865008      UTY       UTY     NR_047643  TSS25449          \n",
       "865014      UTY       UTY  NM_001258256  TSS25449  P29049  \n",
       "\n",
       "[430178 rows x 13 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Filter for only exons\n",
    "genes_df_cp = copy.deepcopy(genes_df)\n",
    "genes_df_cp = genes_df_cp[genes_df_cp[\"feature\"] == \"exon\"]\n",
    "genes_df_cp\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>exon</td>\n",
       "      <td>11874</td>\n",
       "      <td>12227</td>\n",
       "      <td>+</td>\n",
       "      <td>DDX11L1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78313</th>\n",
       "      <td>exon</td>\n",
       "      <td>228337415</td>\n",
       "      <td>228337708</td>\n",
       "      <td>+</td>\n",
       "      <td>GJC2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78307</th>\n",
       "      <td>exon</td>\n",
       "      <td>228336366</td>\n",
       "      <td>228336655</td>\n",
       "      <td>+</td>\n",
       "      <td>GUK1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78314</th>\n",
       "      <td>exon</td>\n",
       "      <td>228345441</td>\n",
       "      <td>228347527</td>\n",
       "      <td>+</td>\n",
       "      <td>GJC2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78306</th>\n",
       "      <td>exon</td>\n",
       "      <td>228336366</td>\n",
       "      <td>228336655</td>\n",
       "      <td>+</td>\n",
       "      <td>GUK1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>865011</th>\n",
       "      <td>exon</td>\n",
       "      <td>15417918</td>\n",
       "      <td>15417992</td>\n",
       "      <td>-</td>\n",
       "      <td>UTY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>865010</th>\n",
       "      <td>exon</td>\n",
       "      <td>15417918</td>\n",
       "      <td>15417992</td>\n",
       "      <td>-</td>\n",
       "      <td>UTY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>865009</th>\n",
       "      <td>exon</td>\n",
       "      <td>15417918</td>\n",
       "      <td>15417992</td>\n",
       "      <td>-</td>\n",
       "      <td>UTY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>865008</th>\n",
       "      <td>exon</td>\n",
       "      <td>15417918</td>\n",
       "      <td>15417992</td>\n",
       "      <td>-</td>\n",
       "      <td>UTY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>865014</th>\n",
       "      <td>exon</td>\n",
       "      <td>15417918</td>\n",
       "      <td>15417992</td>\n",
       "      <td>-</td>\n",
       "      <td>UTY</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>430178 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       feature      start        end strand  gene_id\n",
       "0         exon      11874      12227      +  DDX11L1\n",
       "78313     exon  228337415  228337708      +     GJC2\n",
       "78307     exon  228336366  228336655      +     GUK1\n",
       "78314     exon  228345441  228347527      +     GJC2\n",
       "78306     exon  228336366  228336655      +     GUK1\n",
       "...        ...        ...        ...    ...      ...\n",
       "865011    exon   15417918   15417992      -      UTY\n",
       "865010    exon   15417918   15417992      -      UTY\n",
       "865009    exon   15417918   15417992      -      UTY\n",
       "865008    exon   15417918   15417992      -      UTY\n",
       "865014    exon   15417918   15417992      -      UTY\n",
       "\n",
       "[430178 rows x 5 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Drop unnecessary columns\n",
    "genes_df_cp = genes_df_cp.drop(columns=[\"seqname\", \"source\", \"score\", \"frame\", \"gene_name\", \"transcript_id\", \"tss_id\", \"p_id\"])\n",
    "genes_df_cp\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Since we have duplicate rows, we need to remove them"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>421189</th>\n",
       "      <td>exon</td>\n",
       "      <td>58858172</td>\n",
       "      <td>58858395</td>\n",
       "      <td>-</td>\n",
       "      <td>A1BG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>421193</th>\n",
       "      <td>exon</td>\n",
       "      <td>58858719</td>\n",
       "      <td>58859006</td>\n",
       "      <td>-</td>\n",
       "      <td>A1BG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>421195</th>\n",
       "      <td>exon</td>\n",
       "      <td>58861736</td>\n",
       "      <td>58862017</td>\n",
       "      <td>-</td>\n",
       "      <td>A1BG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>421197</th>\n",
       "      <td>exon</td>\n",
       "      <td>58862757</td>\n",
       "      <td>58863053</td>\n",
       "      <td>-</td>\n",
       "      <td>A1BG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>421200</th>\n",
       "      <td>exon</td>\n",
       "      <td>58863649</td>\n",
       "      <td>58863921</td>\n",
       "      <td>-</td>\n",
       "      <td>A1BG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32957</th>\n",
       "      <td>exon</td>\n",
       "      <td>78097535</td>\n",
       "      <td>78099090</td>\n",
       "      <td>-</td>\n",
       "      <td>ZZZ3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32959</th>\n",
       "      <td>exon</td>\n",
       "      <td>78105134</td>\n",
       "      <td>78105287</td>\n",
       "      <td>-</td>\n",
       "      <td>ZZZ3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32960</th>\n",
       "      <td>exon</td>\n",
       "      <td>78107069</td>\n",
       "      <td>78107131</td>\n",
       "      <td>-</td>\n",
       "      <td>ZZZ3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32961</th>\n",
       "      <td>exon</td>\n",
       "      <td>78107207</td>\n",
       "      <td>78107340</td>\n",
       "      <td>-</td>\n",
       "      <td>ZZZ3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32962</th>\n",
       "      <td>exon</td>\n",
       "      <td>78148270</td>\n",
       "      <td>78148343</td>\n",
       "      <td>-</td>\n",
       "      <td>ZZZ3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>244314 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       feature     start       end strand gene_id\n",
       "421189    exon  58858172  58858395      -    A1BG\n",
       "421193    exon  58858719  58859006      -    A1BG\n",
       "421195    exon  58861736  58862017      -    A1BG\n",
       "421197    exon  58862757  58863053      -    A1BG\n",
       "421200    exon  58863649  58863921      -    A1BG\n",
       "...        ...       ...       ...    ...     ...\n",
       "32957     exon  78097535  78099090      -    ZZZ3\n",
       "32959     exon  78105134  78105287      -    ZZZ3\n",
       "32960     exon  78107069  78107131      -    ZZZ3\n",
       "32961     exon  78107207  78107340      -    ZZZ3\n",
       "32962     exon  78148270  78148343      -    ZZZ3\n",
       "\n",
       "[244314 rows x 5 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Dropping duplicates and sorting data for visualization\n",
    "genes_df_cp = genes_df_cp.drop_duplicates()\n",
    "genes_df_cp.sort_values(by=[\"gene_id\", \"start\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>612673</th>\n",
       "      <td>exon</td>\n",
       "      <td>6027</td>\n",
       "      <td>9963</td>\n",
       "      <td>-</td>\n",
       "      <td>YTHDC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>612681</th>\n",
       "      <td>exon</td>\n",
       "      <td>11954</td>\n",
       "      <td>12088</td>\n",
       "      <td>-</td>\n",
       "      <td>YTHDC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>612685</th>\n",
       "      <td>exon</td>\n",
       "      <td>14159</td>\n",
       "      <td>14221</td>\n",
       "      <td>-</td>\n",
       "      <td>YTHDC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>612689</th>\n",
       "      <td>exon</td>\n",
       "      <td>14330</td>\n",
       "      <td>14369</td>\n",
       "      <td>-</td>\n",
       "      <td>YTHDC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>612693</th>\n",
       "      <td>exon</td>\n",
       "      <td>14466</td>\n",
       "      <td>14502</td>\n",
       "      <td>-</td>\n",
       "      <td>YTHDC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>612698</th>\n",
       "      <td>exon</td>\n",
       "      <td>15763</td>\n",
       "      <td>15845</td>\n",
       "      <td>-</td>\n",
       "      <td>YTHDC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>612702</th>\n",
       "      <td>exon</td>\n",
       "      <td>18389</td>\n",
       "      <td>18555</td>\n",
       "      <td>-</td>\n",
       "      <td>YTHDC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>612706</th>\n",
       "      <td>exon</td>\n",
       "      <td>19779</td>\n",
       "      <td>19863</td>\n",
       "      <td>-</td>\n",
       "      <td>YTHDC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>612709</th>\n",
       "      <td>exon</td>\n",
       "      <td>25642</td>\n",
       "      <td>25759</td>\n",
       "      <td>-</td>\n",
       "      <td>YTHDC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>612713</th>\n",
       "      <td>exon</td>\n",
       "      <td>25842</td>\n",
       "      <td>25950</td>\n",
       "      <td>-</td>\n",
       "      <td>YTHDC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>612717</th>\n",
       "      <td>exon</td>\n",
       "      <td>27743</td>\n",
       "      <td>27837</td>\n",
       "      <td>-</td>\n",
       "      <td>YTHDC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>612720</th>\n",
       "      <td>exon</td>\n",
       "      <td>28434</td>\n",
       "      <td>28487</td>\n",
       "      <td>-</td>\n",
       "      <td>YTHDC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>612723</th>\n",
       "      <td>exon</td>\n",
       "      <td>28948</td>\n",
       "      <td>29037</td>\n",
       "      <td>-</td>\n",
       "      <td>YTHDC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>612727</th>\n",
       "      <td>exon</td>\n",
       "      <td>32667</td>\n",
       "      <td>33090</td>\n",
       "      <td>-</td>\n",
       "      <td>YTHDC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>612732</th>\n",
       "      <td>exon</td>\n",
       "      <td>33212</td>\n",
       "      <td>33540</td>\n",
       "      <td>-</td>\n",
       "      <td>YTHDC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>612736</th>\n",
       "      <td>exon</td>\n",
       "      <td>33923</td>\n",
       "      <td>34024</td>\n",
       "      <td>-</td>\n",
       "      <td>YTHDC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>612739</th>\n",
       "      <td>exon</td>\n",
       "      <td>45366</td>\n",
       "      <td>45746</td>\n",
       "      <td>-</td>\n",
       "      <td>YTHDC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>592623</th>\n",
       "      <td>exon</td>\n",
       "      <td>69176105</td>\n",
       "      <td>69180041</td>\n",
       "      <td>-</td>\n",
       "      <td>YTHDC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>592631</th>\n",
       "      <td>exon</td>\n",
       "      <td>69182032</td>\n",
       "      <td>69182166</td>\n",
       "      <td>-</td>\n",
       "      <td>YTHDC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>592635</th>\n",
       "      <td>exon</td>\n",
       "      <td>69184237</td>\n",
       "      <td>69184299</td>\n",
       "      <td>-</td>\n",
       "      <td>YTHDC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>592639</th>\n",
       "      <td>exon</td>\n",
       "      <td>69184408</td>\n",
       "      <td>69184447</td>\n",
       "      <td>-</td>\n",
       "      <td>YTHDC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>592643</th>\n",
       "      <td>exon</td>\n",
       "      <td>69184544</td>\n",
       "      <td>69184580</td>\n",
       "      <td>-</td>\n",
       "      <td>YTHDC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>592647</th>\n",
       "      <td>exon</td>\n",
       "      <td>69185841</td>\n",
       "      <td>69185923</td>\n",
       "      <td>-</td>\n",
       "      <td>YTHDC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>592651</th>\n",
       "      <td>exon</td>\n",
       "      <td>69188467</td>\n",
       "      <td>69188633</td>\n",
       "      <td>-</td>\n",
       "      <td>YTHDC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>592655</th>\n",
       "      <td>exon</td>\n",
       "      <td>69189857</td>\n",
       "      <td>69189941</td>\n",
       "      <td>-</td>\n",
       "      <td>YTHDC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>592659</th>\n",
       "      <td>exon</td>\n",
       "      <td>69195720</td>\n",
       "      <td>69195837</td>\n",
       "      <td>-</td>\n",
       "      <td>YTHDC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>592663</th>\n",
       "      <td>exon</td>\n",
       "      <td>69195920</td>\n",
       "      <td>69196028</td>\n",
       "      <td>-</td>\n",
       "      <td>YTHDC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>592667</th>\n",
       "      <td>exon</td>\n",
       "      <td>69197821</td>\n",
       "      <td>69197915</td>\n",
       "      <td>-</td>\n",
       "      <td>YTHDC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>592670</th>\n",
       "      <td>exon</td>\n",
       "      <td>69198512</td>\n",
       "      <td>69198565</td>\n",
       "      <td>-</td>\n",
       "      <td>YTHDC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>592673</th>\n",
       "      <td>exon</td>\n",
       "      <td>69199026</td>\n",
       "      <td>69199115</td>\n",
       "      <td>-</td>\n",
       "      <td>YTHDC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>592677</th>\n",
       "      <td>exon</td>\n",
       "      <td>69202745</td>\n",
       "      <td>69203168</td>\n",
       "      <td>-</td>\n",
       "      <td>YTHDC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>592681</th>\n",
       "      <td>exon</td>\n",
       "      <td>69203290</td>\n",
       "      <td>69203618</td>\n",
       "      <td>-</td>\n",
       "      <td>YTHDC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>592686</th>\n",
       "      <td>exon</td>\n",
       "      <td>69204001</td>\n",
       "      <td>69204102</td>\n",
       "      <td>-</td>\n",
       "      <td>YTHDC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>592690</th>\n",
       "      <td>exon</td>\n",
       "      <td>69215444</td>\n",
       "      <td>69215824</td>\n",
       "      <td>-</td>\n",
       "      <td>YTHDC1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       feature     start       end strand gene_id\n",
       "612673    exon      6027      9963      -  YTHDC1\n",
       "612681    exon     11954     12088      -  YTHDC1\n",
       "612685    exon     14159     14221      -  YTHDC1\n",
       "612689    exon     14330     14369      -  YTHDC1\n",
       "612693    exon     14466     14502      -  YTHDC1\n",
       "612698    exon     15763     15845      -  YTHDC1\n",
       "612702    exon     18389     18555      -  YTHDC1\n",
       "612706    exon     19779     19863      -  YTHDC1\n",
       "612709    exon     25642     25759      -  YTHDC1\n",
       "612713    exon     25842     25950      -  YTHDC1\n",
       "612717    exon     27743     27837      -  YTHDC1\n",
       "612720    exon     28434     28487      -  YTHDC1\n",
       "612723    exon     28948     29037      -  YTHDC1\n",
       "612727    exon     32667     33090      -  YTHDC1\n",
       "612732    exon     33212     33540      -  YTHDC1\n",
       "612736    exon     33923     34024      -  YTHDC1\n",
       "612739    exon     45366     45746      -  YTHDC1\n",
       "592623    exon  69176105  69180041      -  YTHDC1\n",
       "592631    exon  69182032  69182166      -  YTHDC1\n",
       "592635    exon  69184237  69184299      -  YTHDC1\n",
       "592639    exon  69184408  69184447      -  YTHDC1\n",
       "592643    exon  69184544  69184580      -  YTHDC1\n",
       "592647    exon  69185841  69185923      -  YTHDC1\n",
       "592651    exon  69188467  69188633      -  YTHDC1\n",
       "592655    exon  69189857  69189941      -  YTHDC1\n",
       "592659    exon  69195720  69195837      -  YTHDC1\n",
       "592663    exon  69195920  69196028      -  YTHDC1\n",
       "592667    exon  69197821  69197915      -  YTHDC1\n",
       "592670    exon  69198512  69198565      -  YTHDC1\n",
       "592673    exon  69199026  69199115      -  YTHDC1\n",
       "592677    exon  69202745  69203168      -  YTHDC1\n",
       "592681    exon  69203290  69203618      -  YTHDC1\n",
       "592686    exon  69204001  69204102      -  YTHDC1\n",
       "592690    exon  69215444  69215824      -  YTHDC1"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Just looking at a single gene's exons to double check\n",
    "genes_df_cp[genes_df_cp[\"gene_id\"] == \"YTHDC1\"].sort_values(by=\"start\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>421198</th>\n",
       "      <td>exon</td>\n",
       "      <td>58863336</td>\n",
       "      <td>58864410</td>\n",
       "      <td>+</td>\n",
       "      <td>A1BG-AS1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>421205</th>\n",
       "      <td>exon</td>\n",
       "      <td>58864745</td>\n",
       "      <td>58864840</td>\n",
       "      <td>+</td>\n",
       "      <td>A1BG-AS1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>421212</th>\n",
       "      <td>exon</td>\n",
       "      <td>58865735</td>\n",
       "      <td>58866549</td>\n",
       "      <td>+</td>\n",
       "      <td>A1BG-AS1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>421209</th>\n",
       "      <td>exon</td>\n",
       "      <td>58865080</td>\n",
       "      <td>58865223</td>\n",
       "      <td>+</td>\n",
       "      <td>A1BG-AS1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174269</th>\n",
       "      <td>exon</td>\n",
       "      <td>9218752</td>\n",
       "      <td>9220651</td>\n",
       "      <td>+</td>\n",
       "      <td>A2M-AS1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>762845</th>\n",
       "      <td>exon</td>\n",
       "      <td>143085860</td>\n",
       "      <td>143086038</td>\n",
       "      <td>+</td>\n",
       "      <td>ZYX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>762829</th>\n",
       "      <td>exon</td>\n",
       "      <td>143079686</td>\n",
       "      <td>143079778</td>\n",
       "      <td>+</td>\n",
       "      <td>ZYX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>762833</th>\n",
       "      <td>exon</td>\n",
       "      <td>143079894</td>\n",
       "      <td>143080415</td>\n",
       "      <td>+</td>\n",
       "      <td>ZYX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>762837</th>\n",
       "      <td>exon</td>\n",
       "      <td>143085312</td>\n",
       "      <td>143085432</td>\n",
       "      <td>+</td>\n",
       "      <td>ZYX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>762841</th>\n",
       "      <td>exon</td>\n",
       "      <td>143085582</td>\n",
       "      <td>143085751</td>\n",
       "      <td>+</td>\n",
       "      <td>ZYX</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>123184 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       feature      start        end strand   gene_id\n",
       "421198    exon   58863336   58864410      +  A1BG-AS1\n",
       "421205    exon   58864745   58864840      +  A1BG-AS1\n",
       "421212    exon   58865735   58866549      +  A1BG-AS1\n",
       "421209    exon   58865080   58865223      +  A1BG-AS1\n",
       "174269    exon    9218752    9220651      +   A2M-AS1\n",
       "...        ...        ...        ...    ...       ...\n",
       "762845    exon  143085860  143086038      +       ZYX\n",
       "762829    exon  143079686  143079778      +       ZYX\n",
       "762833    exon  143079894  143080415      +       ZYX\n",
       "762837    exon  143085312  143085432      +       ZYX\n",
       "762841    exon  143085582  143085751      +       ZYX\n",
       "\n",
       "[123184 rows x 5 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Verify positive strands are there as well\n",
    "genes_df_cp[genes_df_cp[\"strand\"] == \"+\"].sort_values(by=\"gene_id\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Begin introns analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Obtain all unique genes\n",
    "unique_genes = genes_df_cp[\"gene_id\"].unique().tolist()\n",
    "unique_genes.sort()\n",
    "unique_genes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Identify introns for every gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#This may take a few minutes\n",
    "df_lst = []\n",
    "gene_starts = genes_df[genes_df[\"feature\"] == \"start_codon\"].reset_index().drop(columns=[\"index\"])\n",
    "gene_stops = genes_df[genes_df[\"feature\"] == \"stop_codon\"].reset_index().drop(columns=[\"index\"])\n",
    "\n",
    "for gene in unique_genes:\n",
    "    print(gene)\n",
    "    one_gene_df = genes_df_cp[genes_df_cp[\"gene_id\"] == gene].sort_values(by=\"start\")\n",
    "    \n",
    "    if len(one_gene_df) > 1:\n",
    "        adj_df = one_gene_df.sort_values(by=[\"gene_id\", \"start\"]).drop(index=one_gene_df.index[0], axis=0, inplace=False)\n",
    "\n",
    "        new_starts = adj_df[\"start\"].to_list()\n",
    "        if not gene_stops[gene_stops[\"gene_id\"] == gene].empty and gene_stops[gene_stops[\"gene_id\"] == gene][\"start\"].iloc[0] > max(one_gene_df[\"end\"]):\n",
    "            new_starts.append(gene_stops[gene_stops[\"gene_id\"] == gene][\"start\"].iloc[0])\n",
    "        else:\n",
    "            one_gene_df.drop(one_gene_df.tail(1).index,inplace=True)\n",
    "\n",
    "        one_gene_df[\"Next_exon_start\"] = new_starts\n",
    "        #Remove exons that have the same start and next exon start\n",
    "        lowest_idx = one_gene_df.index[one_gene_df[\"start\"] == one_gene_df.iloc[-1][\"Next_exon_start\"]].tolist()\n",
    "        if len(lowest_idx) > 0:\n",
    "            for i in lowest_idx:\n",
    "                if one_gene_df.at[i, \"start\"] == one_gene_df.at[i, \"Next_exon_start\"]:\n",
    "                    one_gene_df = one_gene_df.drop([i])\n",
    "        \n",
    "        if len(one_gene_df) > 0:\n",
    "            #If the exon starting positions are the same for the last exons\n",
    "            last_next_exon = one_gene_df.iloc[-1][\"Next_exon_start\"]\n",
    "            last_exon_starts_lst = one_gene_df.index[one_gene_df[\"start\"] == one_gene_df.iloc[-1][\"start\"]].tolist()\n",
    "            if last_exon_starts_lst[-1] == one_gene_df.last_valid_index():\n",
    "                last_exon_starts_lst.pop()\n",
    "            if len(last_exon_starts_lst) > 0:\n",
    "                for idx in last_exon_starts_lst:\n",
    "                    one_gene_df.at[idx, \"Next_exon_start\"] = last_next_exon\n",
    "\n",
    "            #one_gene_df[\"Intron_start\"] = one_gene_df[\"Next_exon_start\"] - abs(one_gene_df[\"end\"] - one_gene_df[\"Next_exon_start\"]) + 1 \n",
    "            one_gene_df[\"Intron_start\"] = one_gene_df[\"end\"] + 1\n",
    "            one_gene_df[\"Intron_stop\"] = one_gene_df[\"Next_exon_start\"] - 1\n",
    "            one_gene_df = one_gene_df.reset_index().drop(columns=[\"index\"])\n",
    "\n",
    "            if not gene_starts[gene_starts[\"gene_id\"] == gene][\"start\"].empty:\n",
    "                if min(one_gene_df[\"start\"]) != gene_starts[gene_starts[\"gene_id\"] == gene][\"start\"].iloc[0] and gene_starts[gene_starts[\"gene_id\"] == gene][\"start\"].iloc[0] < min(one_gene_df[\"start\"]):\n",
    "                    initial_intron_df = pd.DataFrame({\"feature\": [\"intron\"], \"start\": [np.nan], \"end\": [np.nan],\n",
    "                                                      \"strand\": [one_gene_df[\"strand\"].iloc[0]],\n",
    "                                                      \"gene_id\": [gene], \"Next_exon_start\": [min(one_gene_df[\"start\"])], \n",
    "                                                       \"Intron_start\": [gene_starts[gene_starts[\"gene_id\"] == gene][\"end\"].iloc[0] + 1],\n",
    "                                                      \"Intron_stop\": [min(one_gene_df[\"start\"]) - 1]})\n",
    "                    one_gene_df.append(initial_intron_df)\n",
    "\n",
    "            #Deals with alternatively spliced exons\n",
    "            grouped_df = one_gene_df.groupby(\"start\").mean().reset_index()\n",
    "            same_starts = (one_gene_df[\"start\"] == one_gene_df[\"Next_exon_start\"])\n",
    "            same_starts_df = one_gene_df.loc[same_starts, [\"start\", \"Next_exon_start\"]][\"start\"]\n",
    "            if len(same_starts_df) > 0:\n",
    "                for current_start in list(set(same_starts_df.tolist())):\n",
    "                    for i in one_gene_df.index[one_gene_df[\"start\"] == current_start].tolist():\n",
    "                        if not grouped_df.iloc[grouped_df.index[grouped_df['start'] == current_start].tolist()[0]].equals(grouped_df.iloc[-1]):\n",
    "                            start_replace = grouped_df.iloc[grouped_df.index[grouped_df['start'] == current_start].tolist()[0] + 1][\"start\"]\n",
    "                            one_gene_df.at[i, \"Next_exon_start\"] = start_replace\n",
    "                            one_gene_df.at[i, \"Intron_stop\"] = one_gene_df.iloc[i][\"Next_exon_start\"] - 1\n",
    "\n",
    "            df_lst.append(one_gene_df)\n",
    "    else:\n",
    "        if not gene_starts[gene_starts[\"gene_id\"] == gene].empty:\n",
    "            if one_gene_df[\"start\"].iloc[0] != gene_starts[gene_starts[\"gene_id\"] == gene][\"start\"].iloc[0] and gene_starts[gene_starts[\"gene_id\"] == gene][\"end\"].iloc[0] < one_gene_df[\"start\"].iloc[0]:\n",
    "                initial_intron_df = pd.DataFrame({\"feature\": [\"intron\"], \"start\": [np.nan], \"end\": [np.nan],\n",
    "                                              \"strand\": [one_gene_df[\"strand\"].iloc[0]],\n",
    "                                              \"gene_id\": [gene], \"Next_exon_start\": [min(one_gene_df[\"start\"])], \n",
    "                                               \"Intron_start\": [gene_starts[gene_starts[\"gene_id\"] == gene][\"end\"].iloc[0] + 1],\n",
    "                                              \"Intron_stop\": [min(one_gene_df[\"start\"]) - 1]})\n",
    "                df_lst.append(initial_intron_df)\n",
    "        if not gene_stops[gene_stops[\"gene_id\"] == gene].empty:\n",
    "            if one_gene_df[\"end\"].iloc[0] != gene_stops[gene_stops[\"gene_id\"] == gene][\"end\"].iloc[0] and gene_stops[gene_stops[\"gene_id\"] == gene][\"start\"].iloc[0] > one_gene_df[\"end\"].iloc[0]:\n",
    "                initial_intron_df = pd.DataFrame({\"feature\": [\"intron\"], \"start\": [np.nan], \"end\": [np.nan],\n",
    "                                                  \"strand\": [one_gene_df[\"strand\"].iloc[0]],\n",
    "                                                  \"gene_id\": [gene], \"Next_exon_start\": [gene_stops[gene_stops[\"gene_id\"] == gene][\"start\"]].iloc[0], \n",
    "                                                   \"Intron_start\": [max(one_gene_df[\"start\"]) + 1],\n",
    "                                                  \"Intron_stop\": [gene_stops[gene_stops[\"gene_id\"] == gene][\"start\"].iloc[0] - 1]})\n",
    "                df_lst.append(initial_intron_df)\n",
    "        \n",
    "    \n",
    "#Uncomment if introns are needed when end positions are the same######################################################\n",
    "#    #Obtains duplicated end values\n",
    "#     same_ends_df = one_gene_df.loc[one_gene_df[one_gene_df[\"gene_id\"] == gene].duplicated(subset=[\"end\"], keep=False)]\n",
    "#     for unique_end in same_ends_df[\"end\"].unique().tolist():\n",
    "#         #Subset df for each same end position\n",
    "#         tmp2 = one_gene_df[one_gene_df[\"end\"] == unique_end]\n",
    "#         #ID first instance of preceding value\n",
    "#         preceding_row = one_gene_df.iloc[[tmp2.index[0] - 1]]\n",
    "#         if 0 not in tmp2.index.tolist():\n",
    "#             #Drop first because we already have it\n",
    "#             tmp2.drop(tmp2.index[0], inplace=True)\n",
    "#             #Add preceding row to df\n",
    "#             tmp2.append(preceding_row)\n",
    "#             #Sort values by start so preceding row is on top\n",
    "#             tmp2 = tmp2.sort_values(by=\"start\")\n",
    "#             #Get new intron start and end positions\n",
    "#             intron_end_lst = [start-1 for start in tmp2[\"start\"].tolist()]\n",
    "#             end_length = len(intron_end_lst)\n",
    "#             intron_start_lst = [preceding_row[\"Intron_start\"].iloc[0]] * end_length\n",
    "#             add_df = pd.DataFrame({\"feature\": [np.nan]*end_length, \"start\": [np.nan]*end_length, \"end\": [np.nan]*end_length,\n",
    "#                                                       \"strand\": [np.nan]*end_length,\n",
    "#                                                       \"gene_id\": [gene]*end_length, \"Next_exon_start\": [np.nan]*end_length, \n",
    "#                                                        \"Intron_start\": intron_start_lst,\n",
    "#                                                       \"Intron_stop\": intron_end_lst})\n",
    "#             df_lst.append(add_df)\n",
    "######################################################################################################################   \n",
    "    \n",
    "exons_introns_df = pd.concat(df_lst)\n",
    "index_names = (exons_introns_df[\"Intron_start\"] < exons_introns_df[\"Intron_stop\"])\n",
    "exons_introns_df = exons_introns_df.loc[index_names]\n",
    "exons_introns_df\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>DDX11L1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78313</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>GJC2</td>\n",
       "      <td>GJC2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78307</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>GUK1</td>\n",
       "      <td>GUK1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78318</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>IBA57</td>\n",
       "      <td>IBA57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78334</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>OBSCN</td>\n",
       "      <td>OBSCN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>863499</th>\n",
       "      <td>chrY</td>\n",
       "      <td>unknown</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>CD99P1</td>\n",
       "      <td>CD99P1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>863504</th>\n",
       "      <td>chrY</td>\n",
       "      <td>unknown</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>CD99</td>\n",
       "      <td>CD99</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>863548</th>\n",
       "      <td>chrY</td>\n",
       "      <td>unknown</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>XGPY2</td>\n",
       "      <td>XGPY2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>863454</th>\n",
       "      <td>chrY</td>\n",
       "      <td>unknown</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>DHRSX</td>\n",
       "      <td>DHRSX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>863467</th>\n",
       "      <td>chrY</td>\n",
       "      <td>unknown</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ZBED1</td>\n",
       "      <td>ZBED1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>25266 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       seqname   source strand frame  gene_id gene_name\n",
       "0         chr1  unknown      +     .  DDX11L1   DDX11L1\n",
       "78313     chr1  unknown      +     .     GJC2      GJC2\n",
       "78307     chr1  unknown      +     .     GUK1      GUK1\n",
       "78318     chr1  unknown      +     .    IBA57     IBA57\n",
       "78334     chr1  unknown      +     .    OBSCN     OBSCN\n",
       "...        ...      ...    ...   ...      ...       ...\n",
       "863499    chrY  unknown      +     .   CD99P1    CD99P1\n",
       "863504    chrY  unknown      +     .     CD99      CD99\n",
       "863548    chrY  unknown      +     .    XGPY2     XGPY2\n",
       "863454    chrY  unknown      -     .    DHRSX     DHRSX\n",
       "863467    chrY  unknown      -     .    ZBED1     ZBED1\n",
       "\n",
       "[25266 rows x 6 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Merge dataframes containing all the necessary information to create a gtf file\n",
    "to_merge_df = genes_df[[\"seqname\", \"source\", \"feature\", \"strand\", \"frame\", \"gene_id\", \"gene_name\"]]\n",
    "to_merge_df = to_merge_df[to_merge_df[\"feature\"] == \"exon\"].drop(columns=\"feature\")\n",
    "to_merge_df = to_merge_df.drop_duplicates()\n",
    "to_merge_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>tss_id</th>\n",
       "      <th>p_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr19</td>\n",
       "      <td>unknown</td>\n",
       "      <td>intron</td>\n",
       "      <td>58858396</td>\n",
       "      <td>58858718</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>A1BG</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr19</td>\n",
       "      <td>unknown</td>\n",
       "      <td>intron</td>\n",
       "      <td>58859007</td>\n",
       "      <td>58861735</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>A1BG</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr19</td>\n",
       "      <td>unknown</td>\n",
       "      <td>intron</td>\n",
       "      <td>58862018</td>\n",
       "      <td>58862756</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>A1BG</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr19</td>\n",
       "      <td>unknown</td>\n",
       "      <td>intron</td>\n",
       "      <td>58863054</td>\n",
       "      <td>58863648</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>A1BG</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr19</td>\n",
       "      <td>unknown</td>\n",
       "      <td>intron</td>\n",
       "      <td>58863922</td>\n",
       "      <td>58864293</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>A1BG</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>286152</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>intron</td>\n",
       "      <td>78050341</td>\n",
       "      <td>78097534</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>ZZZ3</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>286153</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>intron</td>\n",
       "      <td>78099091</td>\n",
       "      <td>78105133</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>ZZZ3</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>286154</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>intron</td>\n",
       "      <td>78105288</td>\n",
       "      <td>78107068</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>ZZZ3</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>286155</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>intron</td>\n",
       "      <td>78107132</td>\n",
       "      <td>78107206</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>ZZZ3</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>286156</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>intron</td>\n",
       "      <td>78107341</td>\n",
       "      <td>78148269</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>ZZZ3</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>286157 rows Ã— 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       seqname   source feature     start       end score strand frame  \\\n",
       "0        chr19  unknown  intron  58858396  58858718     .      -     .   \n",
       "1        chr19  unknown  intron  58859007  58861735     .      -     .   \n",
       "2        chr19  unknown  intron  58862018  58862756     .      -     .   \n",
       "3        chr19  unknown  intron  58863054  58863648     .      -     .   \n",
       "4        chr19  unknown  intron  58863922  58864293     .      -     .   \n",
       "...        ...      ...     ...       ...       ...   ...    ...   ...   \n",
       "286152    chr1  unknown  intron  78050341  78097534     .      -     .   \n",
       "286153    chr1  unknown  intron  78099091  78105133     .      -     .   \n",
       "286154    chr1  unknown  intron  78105288  78107068     .      -     .   \n",
       "286155    chr1  unknown  intron  78107132  78107206     .      -     .   \n",
       "286156    chr1  unknown  intron  78107341  78148269     .      -     .   \n",
       "\n",
       "       gene_id gene_name transcript_id tss_id p_id  \n",
       "0         A1BG      A1BG                            \n",
       "1         A1BG      A1BG                            \n",
       "2         A1BG      A1BG                            \n",
       "3         A1BG      A1BG                            \n",
       "4         A1BG      A1BG                            \n",
       "...        ...       ...           ...    ...  ...  \n",
       "286152    ZZZ3      ZZZ3                            \n",
       "286153    ZZZ3      ZZZ3                            \n",
       "286154    ZZZ3      ZZZ3                            \n",
       "286155    ZZZ3      ZZZ3                            \n",
       "286156    ZZZ3      ZZZ3                            \n",
       "\n",
       "[286157 rows x 13 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Create dataframe only containing introns and format as a gtf file\n",
    "introns_df = pd.DataFrame({\"start\": exons_introns_df[\"Intron_start\"].tolist(), \"end\": exons_introns_df[\"Intron_stop\"].tolist(), \n",
    "                          \"gene_id\": exons_introns_df[\"gene_id\"].tolist()})\n",
    "introns_df[\"feature\"] = \"intron\"\n",
    "introns_df[\"score\"] = \".\"\n",
    "introns_df[\"transcript_id\"] = np.nan\n",
    "introns_df[\"tss_id\"] = np.nan\n",
    "introns_df[\"p_id\"] = np.nan\n",
    "introns_df = introns_df.merge(to_merge_df, on=\"gene_id\", how=\"left\")\n",
    "introns_df = introns_df[[\"seqname\", \"source\", \"feature\", \"start\", \"end\", \"score\", \"strand\", \"frame\", \"gene_id\",\n",
    "                       \"gene_name\", \"transcript_id\", \"tss_id\", \"p_id\"]]\n",
    "introns_df[\"frame\"] = \".\"\n",
    "introns_df.replace(np.nan, \"\", inplace=True)\n",
    "introns_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>tss_id</th>\n",
       "      <th>p_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>exon</td>\n",
       "      <td>11874</td>\n",
       "      <td>12227</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>NR_046018</td>\n",
       "      <td>TSS14844</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75545</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>CDS</td>\n",
       "      <td>216497510</td>\n",
       "      <td>216497694</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>USH2A</td>\n",
       "      <td>USH2A</td>\n",
       "      <td>NM_206933</td>\n",
       "      <td>TSS27031</td>\n",
       "      <td>P4249</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75548</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>CDS</td>\n",
       "      <td>216498647</td>\n",
       "      <td>216498941</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>USH2A</td>\n",
       "      <td>USH2A</td>\n",
       "      <td>NM_007123</td>\n",
       "      <td>TSS27031</td>\n",
       "      <td>P5706</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75549</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>CDS</td>\n",
       "      <td>216498647</td>\n",
       "      <td>216498941</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>USH2A</td>\n",
       "      <td>USH2A</td>\n",
       "      <td>NM_206933</td>\n",
       "      <td>TSS27031</td>\n",
       "      <td>P4249</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75552</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>CDS</td>\n",
       "      <td>216500933</td>\n",
       "      <td>216500996</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>USH2A</td>\n",
       "      <td>USH2A</td>\n",
       "      <td>NM_007123</td>\n",
       "      <td>TSS27031</td>\n",
       "      <td>P5706</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>286152</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>intron</td>\n",
       "      <td>78050341</td>\n",
       "      <td>78097534</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>ZZZ3</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>286153</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>intron</td>\n",
       "      <td>78099091</td>\n",
       "      <td>78105133</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>ZZZ3</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>286154</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>intron</td>\n",
       "      <td>78105288</td>\n",
       "      <td>78107068</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>ZZZ3</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>286155</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>intron</td>\n",
       "      <td>78107132</td>\n",
       "      <td>78107206</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>ZZZ3</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>286156</th>\n",
       "      <td>chr1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>intron</td>\n",
       "      <td>78107341</td>\n",
       "      <td>78148269</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>ZZZ3</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1155361 rows Ã— 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       seqname   source feature      start        end score strand frame  \\\n",
       "0         chr1  unknown    exon      11874      12227     .      +     .   \n",
       "75545     chr1  unknown     CDS  216497510  216497694     .      -     0   \n",
       "75548     chr1  unknown     CDS  216498647  216498941     .      -     1   \n",
       "75549     chr1  unknown     CDS  216498647  216498941     .      -     1   \n",
       "75552     chr1  unknown     CDS  216500933  216500996     .      -     2   \n",
       "...        ...      ...     ...        ...        ...   ...    ...   ...   \n",
       "286152    chr1  unknown  intron   78050341   78097534     .      -     .   \n",
       "286153    chr1  unknown  intron   78099091   78105133     .      -     .   \n",
       "286154    chr1  unknown  intron   78105288   78107068     .      -     .   \n",
       "286155    chr1  unknown  intron   78107132   78107206     .      -     .   \n",
       "286156    chr1  unknown  intron   78107341   78148269     .      -     .   \n",
       "\n",
       "        gene_id gene_name transcript_id    tss_id   p_id  \n",
       "0       DDX11L1   DDX11L1     NR_046018  TSS14844         \n",
       "75545     USH2A     USH2A     NM_206933  TSS27031  P4249  \n",
       "75548     USH2A     USH2A     NM_007123  TSS27031  P5706  \n",
       "75549     USH2A     USH2A     NM_206933  TSS27031  P4249  \n",
       "75552     USH2A     USH2A     NM_007123  TSS27031  P5706  \n",
       "...         ...       ...           ...       ...    ...  \n",
       "286152     ZZZ3      ZZZ3                                 \n",
       "286153     ZZZ3      ZZZ3                                 \n",
       "286154     ZZZ3      ZZZ3                                 \n",
       "286155     ZZZ3      ZZZ3                                 \n",
       "286156     ZZZ3      ZZZ3                                 \n",
       "\n",
       "[1155361 rows x 13 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Create the final dataframe\n",
    "final_df = pd.concat([genes_df, introns_df])\n",
    "final_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/sahil/Documents/Rotations/Britt_Glaunsinger/Michaels_project/gtf.py:106: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df['attribute']=\"\"\n",
      "/Users/sahil/Documents/Rotations/Britt_Glaunsinger/Michaels_project/gtf.py:111: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df['attribute']=df['attribute']+c+' \"'+inGTF[c].astype(str)+'\"; '\n",
      "/Users/sahil/Documents/Rotations/Britt_Glaunsinger/Michaels_project/gtf.py:109: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df['attribute']=df['attribute']+c+' \"'+inGTF[c].astype(str)+'\";'\n"
     ]
    }
   ],
   "source": [
    "#Output final dataframe as a gtf file\n",
    "age.writeGTF(final_df,\"./genes_w_introns.gtf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:gtf_formatting]",
   "language": "python",
   "name": "conda-env-gtf_formatting-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
